source/common/tls/context_impl.cc

```diff
@@ -771,7 +780,10 @@ ClientContextImpl::newSsl(const Network::TransportSocketOptionsConstSharedPtr& o
   return ssl_con;
 }
 
-int ClientContextImpl::newSessionKey(SSL_SESSION* session) {
+int ClientContextImpl::newSessionKey([[maybe_unused]] absl::optional<uint64_t> session_tag,
+                                     SSL_SESSION* session) {
+  // TODO(lambdai): save the session according to ``session_tag``.
```

Comment:
What is the value of the session_tag? 

Would it make sense to somehow have the `Upstream::Host` own the session?

Should the cache be shared by all workers, or be per-worker? 

It would probably make sense for the cache size limit to be an amount per upstream host, not just a fixed quantity.

There probably needs to be a way to delete sessions when the associated `Host` is removed (by EDS update, etc).

Reply:
> What is the value of the session_tag?

In this PR I add the sessionTag() into the TransportSocketOptions. 
The goal is to make other transport socket partipate the session_tag. 
For example, if the proxy protocol transport socket is chained in the Upstrean TLS transport socket, 
some setup may want to choose the dest ip or source ip as the key.

Reply:
> Would it make sense to somehow have the Upstream::Host own the session?

I don't want to because
1. this violate the current ownership. Currently it's onwed by tls context. We can change that but probably not for now
2. host can be garbage collected in OriginalDst cluster

Reply:
> Should the cache be shared by all workers, or be per-worker?

As of now it's  shared by all workers. No plan to change, at least in the first implementation.

> It would probably make sense for the cache size limit to be an amount per upstream host, not just a fixed quantity.

Agreed. That would be a challenge to capture the changes of host member, especially for OriginalDst cluster.